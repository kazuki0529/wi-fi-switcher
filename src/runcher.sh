#!/bin/bash

set -eux

###########################################################
# 引数で渡された評価対象の日付が期間内か判定する
#
# @param target 評価対象の日付(yyyy-MM-dd hh:mm:ss)
# @param period_start 開始日(yyyy-MM-dd hh:mm:ss)
# @param period_end 終了日(yyyy-MM-dd hh:mm:ss)
#
# @return 期間内の場合は0を返し、期間外は1を返す
###########################################################
is_within_period() {
  declare -r target=$1
  [[ "${2}" = "" ]] \
    && declare -r period_start='1990-01-01 00:00:00' \
    || declare -r period_start=$2
  [[ "${3}" = "" ]] \
    && declare -r period_end='1990-01-01 00:00:00' \
    || declare -r period_end=$3

  # 評価対象が期間内か判定する
  declare -r start_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_start" +%s))
  declare -r end_difference=$(expr $(date -d "$target" +%s) - $(date -d "$period_end" +%s))
  if [ $start_difference -ge 0 ] && [ $end_difference -le 0 ]; then
    return 0;
  else
    return 1;
  fi
}

declare -r now=$1
declare -r schedule=$2

cd $(dirname $0)

while IFS=, read start end description; do
  # ヘッダ行判定
  [[ "${start}" == "start" ]] && continue

  is_within_period "${now}" "${start}" "${end}" \
    && (yarn cypress run -s ./cypress/integration/PR-500KI-wifi-on.ts || true) \
    &&  exit 0
done < "${schedule}"

yarn cypress run -s ./cypress/integration/PR-500KI-wifi-off.ts
