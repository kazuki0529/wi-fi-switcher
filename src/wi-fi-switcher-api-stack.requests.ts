import { APIGatewayEvent, APIGatewayProxyResult } from 'aws-lambda';
import moment from 'moment';
import RequestUseCase from './usecase/RequestUseCase';

export async function handler(
  event: APIGatewayEvent,
): Promise<APIGatewayProxyResult> {
  const usecase = RequestUseCase();

  const pathPart = event.resource.substring('/v1'.length);

  switch (pathPart) {
    case '/requests': {
      if (event.httpMethod === 'POST') {
        const body = event.body ? JSON.parse(event.body) : undefined;
        if (!body) throw new Error('The Body section is empty.');

        const req = {
          ...body,
          start: moment(body.start),
          end: moment(body.end),
        };

        const data = await usecase.create(req);
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data),
        };
      }
      if (event.httpMethod === 'GET') {
        const data = await usecase.getAll();
        return {
          statusCode: 200,
          headers: {
            'Content-Type': 'application/json; charset=utf-8',
          },
          body: JSON.stringify(data.map(e => ({
            ...e,
            start: e.start.toISOString(),
            end: e.end.toISOString(),
            createdAt: e.createdAt?.toISOString(),
            updatedAt: e.updatedAt?.toISOString(),
          }))),
        };
      }

      throw new Error('An undefined method has been called.');
    };
    case '/requests/{requestId}': {
      if (event.httpMethod !== 'PUT') throw new Error('An undefined method has been called.');

      const id = event.pathParameters ? event.pathParameters.requestId : undefined;
      if (!id) throw new Error('The id has not been specified.');

      const body = event.body ? JSON.parse(event.body) : undefined;
      if (!body) throw new Error('The Body section is empty.');

      const req = {
        ...body,
        start: body.start ? moment(body.start) : undefined,
        end: body.end ? moment(body.end) : undefined,
      };

      const data = await usecase.update(id, req);
      return {
        statusCode: 200,
        headers: {
          'Content-Type': 'application/json; charset=utf-8',
        },
        body: JSON.stringify(data),
      };
    };
    default: {
      throw new Error('An undefined path has been called.');
    };
  }
}
