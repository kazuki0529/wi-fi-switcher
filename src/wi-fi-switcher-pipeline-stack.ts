import * as codecommit from '@aws-cdk/aws-codecommit';
import * as codepipeline from '@aws-cdk/aws-codepipeline';
import { CodeCommitSourceAction } from '@aws-cdk/aws-codepipeline-actions';
import {
  Construct,
  Stage,
  Stack,
  StackProps,
  StageProps,
} from '@aws-cdk/core';
import {
  CdkPipeline,
  SimpleSynthAction,
} from '@aws-cdk/pipelines';
import {
  WiFiSwitcherApiStack,
} from './wi-fi-switcher-api-stack';
import {
  WiFiSwitcherStack,
} from './wi-fi-switcher-stack';


class Application extends Stage {
  constructor(
    scope: Construct,
    id: string,
    props?: StageProps,
  ) {
    super(scope, id, props);

    new WiFiSwitcherApiStack(
      this,
      'wi-fi-switcher-api',
    );
    new WiFiSwitcherStack(
      this,
      'wi-fi-switcher',
    );
  }
}

export class WiFiSwitcherPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const sourceArtifact = new codepipeline.Artifact();
    const cloudAssemblyArtifact = new codepipeline.Artifact();

    const repo = codecommit.Repository.fromRepositoryName(
      this,
      'repo',
      'wi-fi-switcher',
    ) as codecommit.Repository;

    const pipeline = new CdkPipeline(this, 'Pipeline', {
      pipelineName: 'Wi-Fi-SwitcherPipeline',
      cloudAssemblyArtifact,
      sourceAction: new CodeCommitSourceAction ({
        actionName: 'CodeCommit',
        repository: repo,
        branch: 'main',
        output: sourceArtifact,
      }),
      synthAction: SimpleSynthAction.standardYarnSynth({
        sourceArtifact,
        cloudAssemblyArtifact,
        environment: {
          privileged: true,
        },
        synthCommand: 'npx cdk synth',
      }),
    });

    // 開発用のDeploy
    pipeline.addApplicationStage(
      new Application(this, 'staging'),
    );

    // 本番用のDeploy
    pipeline.addApplicationStage(
      new Application(this, 'prod'),
      {
        manualApprovals: true,
      },
    );
  }
}
