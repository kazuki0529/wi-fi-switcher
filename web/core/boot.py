import logging
import os
from datetime import datetime, timedelta, time

import pandas as pd
import streamlit as st

file = '../../data/schedule.csv'
df = pd.read_csv(file)
df['start'] = pd.to_datetime(df['start'], format='%Y-%m-%d %H:%M:%S')
df['end'] = pd.to_datetime(df['end'], format='%Y-%m-%d %H:%M:%S')


# ページ内のタイトル
st.set_page_config(page_title='ゲームプレイ申請', page_icon=':video_game:')
st.title('ゲームプレイ申請')
st.header('新規申請', anchor="new_request")


# 今日どれくらいやる予定か表示する
today = datetime(*datetime.now().timetuple()[:3])
today_request = df.query(f'start >= "{today}" and start < "{today + timedelta(days=1)}"')
if len(today_request) > 0:
    play_sec = today_request.apply(lambda x: (x['end'] - x['start']).total_seconds(), axis=1).sum()
    st.info(f'本日は {time(hour=int(play_sec / 60 / 60), minute=int(play_sec / 60 % 60), second=int(play_sec % 60))} 分のゲームプレイ申請をしています。')

now = datetime.now()
with st.container():
    start = datetime.combine(
        st.date_input(label='開始日', value=now, min_value=now),
        st.time_input(label='開始時間', value=now)
    )

end = start
end = datetime.combine(
    st.date_input(label='終了日', value=end, min_value=start),
    st.time_input(label='終了時間', value=end)
)

# TODO: 23時とかにゲームやらないようにチェックする
if (end - start).total_seconds() <= 2 * 60 * 60 + 10 * 60:  # 2時間 + 10分（バッファ）
    if st.button('申請'):
        df = df.append({
            'start': start,
            'end': end,
            'status': 'Request'
        }, ignore_index=True)
        df.to_csv(file, index=False)
        logging.info(f'申請が行われました。（申請期間：{start} ～ {end}）')
        st.success('申請内容を登録しました。LINEで承認依頼の連絡してください。')
else:
    st.error('一度にプレイできるゲーム時間は2時間です。')

with st.expander("承認済みの申請を表示する"):
    st.dataframe(df.query('status == "Approve"').sort_values('start', ascending=False))


st.header('申請承認', anchor="request_list")
st.subheader('未承認一覧', anchor='approve')

# 未承認の一覧
st.table(df.query('status != "Approve"'))

st.subheader('承認', anchor='approve')

num = st.number_input(label='対象番号', value=(len(df) - 1), min_value=0, max_value=(len(df) - 1))
password = st.text_input(label='承認コード', type='password')
col1, col2 = st.columns(2)

with col1:
    approved = st.button('承認')
with col2:
    rejected = st.button('却下')

try:
    if approved:
        if password == os.environ['APPROVE_CODE']:
            # NOTE: 承認ステータスに変更して保存する
            df.at[num, 'status'] = 'Approve'
            df.to_csv(file, index=False)
            st.success('承認しました。')
        else:
            raise RuntimeError('承認コードが不適切です。この操作は記録されました。')

    if rejected:
        pass
except RuntimeError as e:
    st.error(e)
    logging.error(e)
