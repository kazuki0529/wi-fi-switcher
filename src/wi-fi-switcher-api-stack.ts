import * as apigateway from '@aws-cdk/aws-apigateway';
import * as dynamo from '@aws-cdk/aws-dynamodb';
import * as iam from '@aws-cdk/aws-iam';;
import * as lambda from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import * as logs from '@aws-cdk/aws-logs';
import * as cdk from '@aws-cdk/core';

import { Construct, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';

export type StackStage = 'staging' | 'prod';

export class WiFiSwitcherApiStack extends Stack {
  public readonly table: dynamo.Table;
  public readonly api: apigateway.RestApi;

  constructor(
    scope: Construct, id: string,
    props?: StackProps,
  ) {
    super(scope, id, props);

    // データ格納領域作成
    this.table =new dynamo.Table(this, 'data', {
      partitionKey: {
        name: '_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_s_key',
        type: dynamo.AttributeType.STRING,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: dynamo.BillingMode.PAY_PER_REQUEST,
    });
    this.table.addGlobalSecondaryIndex({
      indexName: 'GSI-1',
      partitionKey: {
        name: '_gsi1_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_gsi1_s_key',
        type: dynamo.AttributeType.STRING,
      },
      projectionType: dynamo.ProjectionType.ALL,
    });

    /**
     * API作成
     */
    const requests = new NodejsFunction(this, 'requests', {
      tracing: lambda.Tracing.ACTIVE,
      runtime: lambda.Runtime.NODEJS_14_X,
      timeout: cdk.Duration.seconds(60),
      memorySize: 2048,
      environment: {
        REGION: this.region,
        TABLE_NAME: this.table.tableName,
      },
      logRetention: logs.RetentionDays.ONE_MONTH,
      role: new iam.Role(this, 'generateBatchLambdaRole', {
        assumedBy: new iam.ServicePrincipal('lambda.amazonaws.com'),
        managedPolicies: [
          iam.ManagedPolicy.fromAwsManagedPolicyName(
            'service-role/AWSLambdaBasicExecutionRole',
          ),
        ],
      }),
    });
    this.table.grantReadWriteData(requests);

    this.api = new apigateway.RestApi(this, 'game-play', {
      policy: new iam.PolicyDocument({
        statements: [
          new iam.PolicyStatement({
            principals: [new iam.AnyPrincipal()],
            effect: iam.Effect.ALLOW,
            actions: ['execute-api:Invoke'],
            resources: ['execute-api:/prod/*'],
          }),
        ],
      }),
      deployOptions: {
        tracingEnabled: true,
        stageName: 'prod',
      },
    });
    const baseApi = this.api.root.addResource('v1');
    const requestsIntegration = new apigateway.LambdaIntegration(requests);
    const requestsApi = baseApi.addResource('requests');
    requestsApi.addMethod('POST', requestsIntegration);
    requestsApi.addMethod('GET', requestsIntegration);

    requestsApi.addResource('{requestId}').addMethod('PUT', requestsIntegration);

    new cdk.CfnOutput(this, 'ApiUrl', { value: this.api.url });
  }
}