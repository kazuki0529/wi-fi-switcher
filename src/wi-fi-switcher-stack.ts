import * as dynamo from '@aws-cdk/aws-dynamodb';
import { Construct, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';

export class WiFiSwitcherStack extends Stack {
  public readonly table: dynamo.Table;

  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // データ格納領域作成
    this.table =new dynamo.Table(this, 'data', {
      partitionKey: {
        name: '_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_s_key',
        type: dynamo.AttributeType.STRING,
      },
      removalPolicy: RemovalPolicy.DESTROY,
      billingMode: dynamo.BillingMode.PAY_PER_REQUEST,
    });
    this.table.addGlobalSecondaryIndex({
      indexName: 'GSI-1',
      partitionKey: {
        name: '_gsi1_p_key',
        type: dynamo.AttributeType.STRING,
      },
      sortKey: {
        name: '_gsi1_s_key',
        type: dynamo.AttributeType.STRING,
      },
      projectionType: dynamo.ProjectionType.ALL,
    });
  }
}