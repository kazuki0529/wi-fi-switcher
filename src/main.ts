import { App } from '@aws-cdk/core';
import { WiFiSwitcherStack } from './wi-fi-switcher-stack';


// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new WiFiSwitcherStack(app, 'wi-fi-switcher-stack', { env: devEnv });

app.synth();
