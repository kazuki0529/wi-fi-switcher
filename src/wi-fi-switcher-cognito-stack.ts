import * as cognito from '@aws-cdk/aws-cognito';
import * as cdk from '@aws-cdk/core';

import { Construct, Stack, StackProps } from '@aws-cdk/core';
import { StackStage } from './wi-fi-switcher-stack';

interface WiFiSwitcherCognitoStackProps extends StackProps {
  readonly stage: StackStage;
}

export class WiFiSwitcherCognitoStack extends Stack {
  readonly userPool: cognito.UserPool;
  readonly userPoolClient: cognito.UserPoolClient;

  constructor(
    scope: Construct, id: string,
    props: WiFiSwitcherCognitoStackProps,
  ) {
    super(scope, id, props);

    const zoneName = this.node.tryGetContext('ZONE_NAME');
    const domain = `${(props.stage === 'staging') ? 'dev.': ''}${zoneName}`;
    const fqdn = `game-play.home.${domain}`;

    this.userPool = new cognito.UserPool(this, 'userPool', {
      selfSignUpEnabled: false,
      standardAttributes: {
        email: { required: true, mutable: true },
        phoneNumber: { required: false },
      },
      signInCaseSensitive: false,
      autoVerify: { email: true },
      signInAliases: { username: true },
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
    this.userPool.addDomain('domain', {
      cognitoDomain: { domainPrefix: fqdn.toLowerCase().split('.').join('-') },
    });

    this.userPoolClient = this.userPool.addClient('client', {
      oAuth: {
        scopes: [
          cognito.OAuthScope.EMAIL,
          cognito.OAuthScope.OPENID,
          cognito.OAuthScope.PROFILE,
        ],
        callbackUrls: [`https://${fqdn}/`],
        logoutUrls: [`https://${fqdn}/`],
        flows: { authorizationCodeGrant: true },
      },
    });
  }
}