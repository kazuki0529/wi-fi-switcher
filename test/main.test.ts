import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { WiFiSwitcherStack } from '../src/wi-fi-switcher-stack';

test('Snapshot', () => {
  const app = new App();
  const stack = new WiFiSwitcherStack(app, 'test');

  expect(stack).toHaveResource('AWS::DynamoDB::Table');
  expect(app.synth().getStackArtifact(stack.artifactId).template).toMatchSnapshot();
});
