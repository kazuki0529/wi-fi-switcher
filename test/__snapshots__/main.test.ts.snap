// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Resources": Object {
    "data7E2128CA": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "p_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "s_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "gsi1_p_key",
            "AttributeType": "S",
          },
          Object {
            "AttributeName": "gsi1_s_key",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": Array [
          Object {
            "IndexName": "GSI-1",
            "KeySchema": Array [
              Object {
                "AttributeName": "gsi1_p_key",
                "KeyType": "HASH",
              },
              Object {
                "AttributeName": "gsi1_s_key",
                "KeyType": "RANGE",
              },
            ],
            "Projection": Object {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": Array [
          Object {
            "AttributeName": "p_key",
            "KeyType": "HASH",
          },
          Object {
            "AttributeName": "s_key",
            "KeyType": "RANGE",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
